### PR TITLE
haskell-ghc: fix darwin issue

### PR DESCRIPTION
**Pull request by kalhauge**
###### Motivation for this change

Cabal does not adequately add include-libs to the DYLD_LIBRARY_PATH environment, which is needed by some of the internals, when using external libraries. This pull-request is a fix to #16357, which now works. 

Downside: This change require recompilations of all libraries, so I have tired to write the change so that it only affects Darwin users. 

Also, ideally this should be fixed from Cabal's side but until then this will keep the wheels spinning.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [x] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
  Z3 now loads from ghci.

```
nix-shell -p "(import ./. {}).haskell.packages.ghc7102.ghcWithPackages (p: [ p.z3 ])"
[nix-shell:~/nixpkgs]$ ghci
GHCi, version 7.10.2: http://www.haskell.org/ghc/  :? for help
Prelude> import Z3.Monad
Prelude Z3.Monad>
```
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/haskell-modules/generic-builder.nix [hunk 1]
```diff
@@ -192,10 +192,12 @@ stdenv.mkDerivation ({
         configureFlags+=" --extra-include-dirs=$p/include"
       fi
       if [ -d "$p/lib" ]; then
-        configureFlags+=" --extra-lib-dirs=$p/lib"
-        if [[ ${ if stdenv.isDarwin then "yes" else "no"} = "yes" ]]; then
-          export DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:$p/lib;
-        fi
+        configureFlags+=" --extra-lib-dirs=$p/lib"${
+          if stdenv.isDarwin then
```
Comment (LnL7) [inline pkgs/development/haskell-modules/generic-builder.nix]: you can use `stdenv.lib.optionalString` for this

File: pkgs/development/haskell-modules/generic-builder.nix [hunk 2]
```diff
@@ -192,10 +192,12 @@ stdenv.mkDerivation ({
...
+            "; export DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:$p/lib;"
```
Comment (LnL7) [inline pkgs/development/haskell-modules/generic-builder.nix]: this looks strange, why are the semi colons there?

**Comment by peti** [comment]
Appending a full-blown shell  command to $configureFlags violates the assumption that this variable contains a list of flags, i.e. the change relies on magic knowledge that $configureFlags is only going to be used in places where this kind of hack actually works. Personally, I'd rather have a clean solution for the problem, even if it means re-compiling all packages.

Also, can you please share a concrete test case how to reproduce the issue this patch fixes? Which build used to fail does now work with the patch applied?

**Comment by LnL7** [comment]
@peti he did include an example: `haskellPackages.z3`

**Comment by kalhauge** [comment]
@peti and @LnL7, The semi-colons are there to separate the commands from each other. If the ox is darwin the command will look like:

```bash
configureFlags+=" --extra-lib-dirs=$p/lib"; export DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:$p/lib;
```

It has therefore nothing to do with configureFlags, and therefor do not violate that assumption (but maybe some other).

The reason why I do this is to interfere as little with no darwin users. The result of those runs will look like this:

```bash
configureFlags+=" --extra-lib-dirs=$p/lib"
```
Which should not change the hash of the build script, so that we do not force non darwin users to recompile all of their libraries.

**Comment by kalhauge** [comment]
@LnL7 the change now uses stdenv.lib.optionalString

**Comment by LnL7** [comment]
@kalhauge you should use a multiline string instead of the semicolons the only thing that impacts linux is the newline before/after the interpolation

**Comment by peti** [comment]
Please use a clean implementation even if it means triggering re-builds.

**Comment by kalhauge** [comment]
Now with a clean implementation :)

**Comment by LnL7** [comment]
I'll need some time to test this.

**Comment by vcunat** [comment]
I suppose this isn't just waiting for a merge to staging?

**Comment by peti** [comment]
Looks good to me. I'd merge, but I'm traveling and don't have commit access right now. If someone else can press the "merge" button for me, please do. :-)

**Pull request merged**